python/ray/ml/storage.py

```diff
@@ -0,0 +1,183 @@
+        )
+
+    def delete(self, remote_target: str) -> None:
+        subprocess.check_call(
+            ["aws", "s3", "rm", "--recursive", "--quiet", remote_target]
+        )
+
+
+class GSStorage(Storage):
+    def upload(self, local_source: str, remote_target: str) -> None:
+        subprocess.check_call(["gsutil", "-m", "cp", "-r", local_source, remote_target])
+
+    def download(self, remote_source: str, local_target: str) -> None:
+        subprocess.check_call(["gsutil", "-m", "cp", "-r", remote_source, local_target])
+
+    def delete(self, remote_target: str) -> None:
+        subprocess.check_call(["gsutil", "-m", "rm", "-f", "-r", remote_target])
+
+
+class HDFSStorage(Storage):
+    def upload(self, local_source: str, remote_target: str) -> None:
+        subprocess.check_call(["hdfs", "dfs", "-put", local_source, remote_target])
+
+    def download(self, remote_source: str, local_target: str) -> None:
+        subprocess.check_call(["hdfs", "dfs", "-get", remote_source, local_target])
+
+    def delete(self, remote_target: str) -> None:
+        subprocess.check_call(["hdfs", "dfs", "-rm", "-r", remote_target])
+
+
+_registered_storages: Dict[str, Storage] = {}
+
+
+@DeveloperAPI
+def get_storage(uri: str) -> Storage:
+    """Get external storage provider for a bucket URI.
+
+    Example:
+
+        storage = get_external_storage("s3://test/bucket")
+        assert isinstance(storage, S3Storage)
+
+
+    Args:
+        uri: Bucket URI, e.g. ``s3://bucket/path``
+
+    Returns: ``ExternalStorage`` class.
+
+    Raises: ValueError if no external storage class is found.
+    """
+    global _registered_storages
+    for prefix, storage in _registered_storages.items():
+        if uri.startswith(prefix):
+            return storage
+    raise ValueError(f"No external storage provider found for URI: {uri}")
+
+
+@DeveloperAPI
+def register_storage(prefix: str, storage: Storage, override: bool = True) -> None:
+    """Register storage provider.
+
+    If a prefix is already registered, it will be overwritten without warning,
+    except when ``override=False``, in which case it is ignored without
+    warning.
+
+    Args:
+        prefix: String prefix to identify storage URI, e.g. ``prefix://``.
+        storage: Storage instance.
+        override: If ``True`` (default), will silently override existing
+            storage providers.
+    """
+    global _registered_storages
+    if not override and prefix in _registered_storages:
+        return
+    _registered_storages[prefix] = storage
+
+
+# Register default storages. Do not override if there are e.g.
+# user-provided overrides for these storages.
+register_storage(S3_PREFIX, S3Storage(), override=False)
+register_storage(GS_PREFIX, GSStorage(), override=False)
+register_storage(HDFS_PREFIX, HDFSStorage(), override=False)
+
+
+def is_cloud_target(target: str):
+    global _registered_storages
+    return any(target.startswith(prefix) for prefix in _registered_storages.keys())
```

Comment:
nit: Seems from this definition that Storage is synonymous with cloud storage, is this intentional?

Reply:
`is_cloud_target` is a legacy function name, but e.g. HDFS is not really cloud storage